### PR TITLE
 Multiple word searching with List.search - closes #680 

### LINKED ISSUE #680: Multiple word searching with List.search
At present `List.search` matches list items against a simple substring, and while `List.fuzzySearch` supports multiple words, it is very slow and seems to be better suited to fuzzy matching larger blocks of text rather than many short strings?

Below is an extension of my faster search function in #679 that now supports multiple-word searches (space-separated words, logical AND) as well as quoted phrases that are matched exactly:
```javascript
listObj.search('Jon 1986', searchFunction); // Custom search for Jon AND 1986
listObj.search('"Jonny S" 1986', searchFunction); // Custom search for "Jonny S" AND 1986
```
note that whitespace is stripped from search words unless contained inside quotes. So `search('" a "')` (matches the separate word 'a') is very different to `search(' a ')` (matches anything with the character 'a').

This search function is faster than the default `List.search` in almost all cases, even searching on a single word, and is always _many_ times faster than `List.fuzzySearch`.
```javascript
// searchString is converted to lowercase by List.js
function searchFunction(searchString, columns){
  // Extract quoted phrases "word1 word2" from original searchString
  var words = [],  phrase;
  while ((phrase = searchString.match(/"([^"]+)"/)) !== null) {
    words.push(phrase[1]);
    searchString = searchString.substring(0,phrase.index) + searchString.substring(phrase.index+phrase[0].length);
  };
  // Get remaining space-separated words (if any)
  searchString = searchString.trim();
  if (searchString.length) words = words.concat(searchString.split(/\s+/));
  for (var k = 0, kl = listObj.items.length; k < kl; k++) {
    var item = listObj.items[k];
    item.found = false;
    if (!words.length) continue;
    var values = item.values();
    for (var i = 0, il = words.length; i < il; i++) {
      var word_found = false;
      for (var j = 0, jl = columns.length; j < jl; j++) {
        var column = columns[j];
        if (values.hasOwnProperty(column) && values[column] !== undefined && values[column] !== null) {
          var text = (typeof values[column] !== 'string') ? values[column].toString() : values[column];
          if (text.toLowerCase().indexOf(words[i]) !== -1) {
            // word found, so no need to check it against any other columns
            word_found = true;
            break;
          }
        }
      }
      // this word not found? no need to check any other words, the item cannot match
      if (!word_found) break;
    }
    item.found = word_found;
  }
};
```

### PR DESCRIPTION
**Pull request by sheffieldnick**
Default search function now supports fast multi-word searching as well as exact matching of quoted phrases. See #680 

Added documentation to List API and test units to `__test__/search.test.js`
(also fixes a few typos in `__test__/search.test.js`)

### CODE REVIEW

**Comment by sheffieldnick** [comment]
If this multi-word search is accepted into the library, replacing the default search, is it worth having fuzzy search built-in as well? Fuzzy search is 10% of the List.min.js size, and it could easily be provided as a separate file `list-fuzzysearch.js` (doesn't even need to be a plugin) that provides a custom search function for projects that need it?

File: src/search.js [hunk 1]
```diff
@@ -45,28 +45,40 @@ module.exports = function(list) {
...
-          return;
-        }
-      }
-    },
-    values: function(values, column) {
-      if (values.hasOwnProperty(column)) {
-        text = list.utils.toString(values[column]).toLowerCase();
-        if ((searchString !== "") && (text.search(searchString) > -1)) {
-          return true;
+        var item = list.items[k];
```
Comment (buczek) [inline src/search.js:59]: `for (var item of list.items) {` ?

File: src/search.js [hunk 2]
```diff
@@ -45,28 +45,40 @@ module.exports = function(list) {
...
+        item.found = false;
+        if (!words.length) continue;
+        for (var i = 0, il = words.length; i < il; i++) {
+          var word_found = false;
+          for (var j = 0, jl = columns.length; j < jl; j++) {
+            var values = item.values(), column = columns[j];
```
Comment (buczek) [inline src/search.js:65]: Lift `var values = item.values()`out of the two inner loops?

File: src/search.js (hunk 1 above)
Comment (sheffieldnick) [reply src/search.js:59]: No. The array iteration used in the code was tested as being the fastest across multiple browsers, and is also compatible with all browsers. `for ... of ...` isn't compatible with older browsers.

File: src/search.js (hunk 2 above)
Comment (sheffieldnick) [reply src/search.js:65]: Yes, definite improvement :) Thanks

File: src/search.js (hunk 1 above)
Comment (buczek) [reply src/search.js:59]: Then its certainly better the way it is 👍

**Comment by javve** [comment]
Merged into #696

**Comment by eng1neer** [comment]
Great update! Looks like it unexpectedly (well, at least to me 🙂) changed the behavior of searches that contain URLs though: previously in version 1.5.0 I could search for "example.com" just fine, and now it seems to split it into two words and it stopped working. Any way to search for an exact match (quoted `"example.com"` does not work either)?

**Comment by sheffieldnick** [comment]
> Great update! Looks like it unexpectedly (well, at least to me slightly_smiling_face) changed the behavior of searches that contain URLs though: previously in version 1.5.0 I could search for "example.com" just fine, and now it seems to split it into two words and it stopped working. Any way to search for an exact match (quoted `"example.com"` does not work either)?

Can you submit a new issue with this bug please? and example code showing it failing. This code _should_ only split words on whitespace (space, tab, newline) characters, not on a period, so 'example.com' should work without needing exact-match quoting.

**Comment by eng1neer** [comment]
@sheffieldnick https://github.com/javve/list.js/issues/715

**Pull request closed without merging**